### DISCUSSION TITLE
How does updateEvent emit boolean when there is a new bar on BarDataList?

### DISCUSSION THREAD

**Discussion by nicholashoyy** [ib-api-reloaded/ib_async#52] [Q&A] [answered]
Hi guys,

I am browsing through source codes for eventkit and ib_async, but still couldn't figure out the relevant codes for returning True value when there is a new bar on a BarDataList (updated by reqHistoricalData with keepUpToDate = True).

In the below example provided by blog AlgoTrading101, there is a callback function on_new_bar to which boolean value & BarDataList are emitted.  I wonder how this boolean value is generated after new bar is received.  I have looked through wrapper.HistoricalDataUpdate, ib.reqHistoriclaData, ib.reqHistoricalDataAsync.  Still no clues.  May I have your kind help?

from ib_insync import *
ib = IB()
ib.connect()

def place_order(direction, qty, df, tp, sl):
    bracket_order = ib.bracketOrder(direction, qty, limitPrice=df.close.iloc[-1], takeProfitPrice=tp, stopLossPrice=sl)

for ord in bracket_order:
    ib.placeOrder(contract, ord)

def on_new_bar(bars: BarDataList, has_new_bar: bool):
    if has_new_bar:
        df = util.df(data)
        sma = df.close.tail(50).mean()
        std_dev = df.close.tail(50).std() * 3

 if contract not in [i.contract for i in ib.positions()]:
       if df.close.iloc[-1] > sma + std_dev:
        place_order('SELL', 1, df, sma, sma +std_dev * 2

 elif df.close.iloc[-1] < sma - std_dev:
       place_order('BUY', 1, df, sma, sma+std_dev * 2)

contract = ContFuture('HE', 'GLOBEX')
ib.qualifyContracts(contract)

data = ib.reqHistoricalData(
    contract,
    endDateTime='',
    durationStr='2 D',
    barSizeSetting='15 mins',
    whatToShow='MIDPOINT',
    useRTH=True,
    keepUpToDate=True,
)

data.updateEvent += on_new_bar

ib.run()

**Comment by jjlinares** [comment, marked as answer]
It's in wrapper.realtimeBar

**Reply by nicholashoyy** [reply]
Thank you very much
